File: typescript/src/api/validation.ts
/**
 * Input validation utilities for the Lightcone REST API.
 */

import { PublicKey } from "@solana/web3.js";
import { ApiError } from "./error";

/** Maximum allowed pagination limit */
export const MAX_PAGINATION_LIMIT = 500;

/** Default request timeout in milliseconds */
export const DEFAULT_TIMEOUT_MS = 30000;

/**
 * Validate a Solana public key (Base58-encoded).
 * Uses @solana/web3.js PublicKey for proper validation.
 * @throws {ApiError} If the value is not a valid Solana pubkey
 */
export function validatePubkey(value: string, fieldName: string): void {
  if (!value || value.length === 0) {
    throw ApiError.invalidParameter(`${fieldName} cannot be empty`);
  }
  try {
    new PublicKey(value);
  } catch {
    throw ApiError.invalidParameter(`${fieldName} is not a valid Solana public key`);
  }
}

/**
 * Validate an Ed25519 signature (128 hex characters).
 * @throws {ApiError} If the signature is invalid
 */
export function validateSignature(signature: string): void {
  if (signature.length !== 128) {
    throw ApiError.invalidParameter(
      `Signature must be 128 hex characters, got ${signature.length}`
    );
  }
  if (!/^[0-9a-fA-F]+$/.test(signature)) {
    throw ApiError.invalidParameter("Signature must contain only hex characters");
  }
}

/**
 * Validate pagination limit (1-500).
 * @throws {ApiError} If the limit is out of bounds
 */
export function validateLimit(limit: number | undefined): void {
  if (limit !== undefined && (limit < 1 || limit > MAX_PAGINATION_LIMIT)) {
    throw ApiError.invalidParameter(`Limit must be 1-${MAX_PAGINATION_LIMIT}`);
  }
}

File: python/src/api/validation.py
"""Input validation utilities for the API client."""

import re

from solders.pubkey import Pubkey

from .error import InvalidParameterError

# Hex signature pattern (128 characters for Ed25519 signature)
HEX_SIGNATURE_PATTERN = re.compile("^[0-9a-fA-F]{128}$")

MAX_PAGINATION_LIMIT = 500


def validate_pubkey(value: str, field_name: str) -> None:
    """Validate that a string is a valid Solana pubkey (Base58).

    Uses solders.Pubkey for proper validation including length check.

    Raises:
        InvalidParameterError: If not a valid pubkey
    """
    if not value or not value.strip():
        raise InvalidParameterError(f"{field_name} cannot be empty")

    try:
        Pubkey.from_string(value)
    except Exception:
        raise InvalidParameterError(f"{field_name} is not a valid pubkey")


def validate_signature(signature: str) -> None:
    """Validate that a signature is 128 hex characters.

    Raises:
        InvalidParameterError: If signature is invalid
    """
    if not HEX_SIGNATURE_PATTERN.match(signature):
        raise InvalidParameterError(
            f"Signature must be 128 hex characters, got {len(signature)}"
        )


def validate_limit(limit: int) -> None:
    """Validate pagination limit is within bounds.

    Raises:
        InvalidParameterError: If limit is out of bounds
    """
    if limit < 1 or limit > MAX_PAGINATION_LIMIT:
        raise InvalidParameterError(f"Limit must be 1-{MAX_PAGINATION_LIMIT}")
